### PR TITLE
Add support for XP progression (#86)

### PR DESCRIPTION
**Pull request by RDIL**
Drafting final merge of #86 into v6 master branch.

### CODE REVIEW

**Comment by LennardF1989** [comment]
Just to elaborate on this, the title doesn't do this PR justice:
- Adds support for Player Profile XP and Mastery XP progression in all gamemodes
- Adds proper score screen on mission end for all gamemodes
    -  [Exception] Sniper missions do not have their score properly calculated yet
- Adds support for Freelancer mastery unlocks
- Adds an extended Player Profile under Career on the main menu
- Adds support for persistent unlocked shortcuts

**Comment by AnthonyFuller** [comment]
> [Exception] Sniper missions do not have their score properly calculated yet

I will add this, in a future PR for a future version of Peacock. v6 is starting to suffer from feature creep.

File: components/candle/challengeService.ts
Comment (AnthonyFuller) [inline components/candle/challengeService.ts:1137]: At the start, the repeat counter starts at whatever the `Base` value is, every time the challenge is completed, the `Delta` value is then taken awake from the "repeat counter". We can probably use base to track this. I presume if the counter is less than 0, it won't activate. These are actually only really used on statistics, the rest of the time the `Base` is 1 with a `Delta` of 0.

File: components/utils.ts
Comment (AnthonyFuller) [inline components/utils.ts:153]: I'll do this, the difference between levels "steadily" increases until it doesn't:
image

File: components/candle/challengeService.ts
Comment (LennardF1989) [reply components/candle/challengeService.ts:1137]: Good to know, thanks! I think it's okay to leave it for now then.

**Review by AnthonyFuller — CHANGES_REQUESTED** [review]
LGTM. Huge thanks for implementing this! Just respond to comments.

File: components/evergreen.ts
Comment (LennardF1989) [reply components/evergreen.ts:57]: I thought about that, but on the Discord we also had talk where it's okay progression happens, as long as people are not locked out of unlocks. Honestly, this was the easy way to go since already had code at that spot initially. In testing, flipping this flag will kind of mess-up Safehouse, since some things will be undone, and some won't.

File: components/flags.ts
Comment (LennardF1989) [reply components/flags.ts:107]: Maybe, I just went with a choice for a player. I can imagine someone might want to play freelancer, but have all doors unlocked without having to play for it.

File: components/profileHandler.ts
Comment (LennardF1989) [reply components/profileHandler.ts:605]: Yes, it depends on the type of contract you are doing. When doing a sniper contract, the sniper LevelInfo is here. The game appears to use it to send "level updated" (I don't have the exact name right now) events during gameplay.

File: static/GlobalChallenges.json
Comment (LennardF1989) [reply static/GlobalChallenges.json:1]: Agreed, I'm happy @moonysolari went through the trouble to do it manually though :P

File: components/evergreen.ts
Comment (AnthonyFuller) [reply components/evergreen.ts:57]: Well at some point in the future, the inventory will be managed by whatever they unlocked through challenges or mastery. So having all this behind a generic "unlock all" flag/button would be ideal even if that's not a feature now.

About flipping it, the idea behind an "unlock all" button would be that the only way to undo it is to reset your profile.

File: components/flags.ts
Comment (AnthonyFuller) [reply components/flags.ts:107]: Couldn't they just go and unlock the shortcuts though?

File: components/scoreHandler.ts
Comment (LennardF1989) [reply components/scoreHandler.ts:81]: Correct, it's just that the current set only has a subset of the ones the Official server hands out. For example, none of the UI_PLAYSTYLE_ICA_* are used. Obviously, needs some imagination on our side to do this.

File: components/scoreHandler.ts
Comment (LennardF1989) [reply components/scoreHandler.ts:217]: Not sure what your question here is referring to. This comment is about merging with the bit in `missionEnd` that handles the non-global challenges: `Object.values(contractChallenges)`. It's almost the same code, so potentially deduplication is possible.

File: components/profileHandler.ts
Comment (AnthonyFuller) [reply components/profileHandler.ts:605]: Okay so some future thing when proper sniper scoring is implemented. Cool. Feel free to resolve.

File: components/scoreHandler.ts
Comment (AnthonyFuller) [reply components/scoreHandler.ts:217]: This is a great question! I remember writing this comment on another line. [This](https://github.com/thepeacockproject/Peacock/blob/4e00104de504950d313fb403adc35776379e8835/components/scoreHandler.ts#L691) is the line I meant to comment on.

File: components/scoreHandler.ts
Comment (LennardF1989) [reply components/scoreHandler.ts:691]: Over the past week as I was testing, sometimes XPGain is different on the different elements of the MissionEnd response, and sometimes (most of the time), they are always the same. "Mastery XP" you earn from mastery challenges are always added, whereas progression wise, all XP goes to both Mastery and Player, so it doesn't make sense...

Luckily, it doesn't seem to matter. The only XP that is actually being looked in the MissionEnd response is the XP under `ProfileProgression`, the others can be literally set to anything.

File: components/scoreHandler.ts
Comment (LennardF1989) [reply components/scoreHandler.ts:861]: I've done a Silent Assassin run on Freelancer, and this was the rank I got at the end. The boolean for silentAssassin is always false in Freelancer, so this was the closest thing possible to make it feel like Official.

File: components/scoreHandler.ts
Comment (AnthonyFuller) [reply components/scoreHandler.ts:81]: Someone is wanting to figure them out. It's going to take a while.

File: components/scoreHandler.ts
Comment (AnthonyFuller) [reply components/scoreHandler.ts:691]: Cool.

File: components/scoreHandler.ts
Comment (LennardF1989) [reply components/scoreHandler.ts:788]: I thought about this bit of a code a long time. The problem is the list of completed objectives reference the "conditionObjective" and potentially the "secondaryObjective" (if it is in state Success).

I have to loop over all GameChangerProperties no matter what to make a map of "this conditionObjective belongs to this secondaryObjective" before I can determine if something is a payout/prestige AND was succesfull/failed.

File: components/scoreHandler.ts
Comment (LennardF1989) [reply components/scoreHandler.ts:1061]: I'll leave that to @RDIL to be the judge off :P

File: components/scoreHandler.ts
Comment (LennardF1989) [reply components/scoreHandler.ts:257]: Maybe, I was going to say sniper contracts do this too. Question is how much of a pain is it going to be to make and maintain that compared to doing it in code...

File: components/scoreHandler.ts
Comment (AnthonyFuller) [reply components/scoreHandler.ts:861]: Fair enough.

File: components/flags.ts
Comment (AnthonyFuller) [reply components/flags.ts:107]: Deferring this to future discussions.

File: components/scoreHandler.ts
Comment (AnthonyFuller) [reply components/scoreHandler.ts:788]: Hmmmm, okay. That's the issue with sending all hundred-and-something gamechangers over.

**Review by AnthonyFuller — APPROVED** [review]
LGTM.

**Comment by AnthonyFuller** [comment]
@RDIL can't request your review. Give it a once over then merge if you see fit.

File: components/scoreHandler.ts
Comment (LennardF1989) [reply components/scoreHandler.ts:788]: It's funny how they decided to do it like this on official, you can notice a significant performance decrease in both GetForPlay2 and MissionEnd for Freelancer.

**Pull request merged**
